### PR TITLE
Fail test when waitFor().withTimeout() reaches timeout and assertion fails

### PR DESCRIPTION
**Pull request by LeoNatan**
Also removed no longer needed comments in docs regarding the behavior of waitFor().

### CODE REVIEW

**Comment by LeoNatan** [comment]
It wasn't "very hard". 😉

**Comment by LeoNatan** [comment]
Related to #1463 , #1471

File: detox/ios/Detox/DTXMethodInvocation.m [hunk 1]
```diff
@@ -15,58 +15,58 @@ @implementation DTXMethodInvocation
...
-    return nil;
+	if (param == nil) return nil;
+	if ([param isKindOfClass:[NSDictionary class]])
+	{
+		NSDictionary *p = (NSDictionary*)param;
+		NSString *type = [DTXMethodInvocation getString:[p objectForKey:@"type"]];
+		id value = [p objectForKey:@"value"];
+		return [DTXMethodInvocation getValue:value withType:type onError:onError];
+	}
+	return nil;
```
Comment (d4vidi) [inline detox/ios/Detox/DTXMethodInvocation.m]: nu ma :-/

Comment (LeoNatan) [reply detox/ios/Detox/DTXMethodInvocation.m]: ?

File: detox/test/e2e/05.waitfor.test.js [hunk 2]
```diff
@@ -38,6 +37,11 @@ describe('WaitFor', () => {
     await expect(element(by.text('Text5'))).toBeVisible();
   });
 
+  it('should fail test after waiting for element to exist but it doesn\'t at the end', async () => {
+    await expect(element(by.id('neverAppearingText'))).toNotExist();
+    await expectToThrow(() => waitFor(element(by.id('neverAppearingText'))).toExist().withTimeout(1000));
```
Comment (d4vidi) [inline detox/test/e2e/05.waitfor.test.js]: I think that equivalently, there should be an `expectToThrow()` over the `should wait until an element is removed` test as well

File: docs/APIRef.waitFor.md [hunk 3]
```diff
@@ -21,60 +21,54 @@ Test async code with waitFor.
 >NOTE: `waitFor` will not throw when reaching timeout, instead it will just continue to the next line. To make sure your tests work as you expect them to add `expect()` at the following line.  
 
 ### `toBeVisible()`
-Test will hang until expectation is met or a timeout has occurred. Should be accompanied with [`expect.toBeVisible()`](APIRef.Expect.md#tobevisible)
+Test will hang until expectation is met or a timeout has occurred.
```
Comment (d4vidi) [inline docs/APIRef.waitFor.md]: kudos for also updating the docs :)

File: detox/ios/Detox/DTXMethodInvocation.m (hunk 1 above)
Comment (d4vidi) [reply detox/ios/Detox/DTXMethodInvocation.m]: spaces => tabs, much pointless diff

File: detox/test/e2e/05.waitfor.test.js [hunk 4]
```diff
@@ -38,6 +37,11 @@ describe('WaitFor', () => {
...
+  });
+
```
Comment (d4vidi) [inline detox/test/e2e/05.waitfor.test.js]: 👍

File: detox/ios/Detox/TestRunner.m [hunk 5]
```diff
@@ -22,59 +23,70 @@ @implementation TestRunner
...
-    });
+	grey_execute_async(^{
+		id res = [DTXMethodInvocation invoke:params onError:^(NSString *error)
+				  {
+					  if (self.delegate) [self.delegate testRunnerOnError:error withMessageId:messageId];
+				  }];
+		
+		NSString* currentSelector = [params objectForKey:@"method"];
+		
+		if([currentSelector isEqualToString:@"waitWithTimeout:"] && [res boolValue] == NO)
```
Comment (d4vidi) [inline detox/ios/Detox/TestRunner.m]: A bit tailored... can't this be generalized to _all_ tests?

Comment (LeoNatan) [reply detox/ios/Detox/TestRunner.m]: I don't know. There is no way to know why a return value is YES/NO, so I didn't feel comfortable making it a general case. It's _probably_ correct to assume of NO then fail, but I can't guarantee.

File: detox/ios/Detox/DTXMethodInvocation.m (hunk 1 above)
Comment (LeoNatan) [reply detox/ios/Detox/DTXMethodInvocation.m]: It's possible to configure this in the diff view:

whitespace

**Pull request merged**